### ISSUE TITLE
Penalty for reopened issues

### ISSUE THREAD

**Issue by rndquu** [ubiquity/ubiquibot#439]
There is an [issue](https://github.com/ubiquity/ubiquity-dollar/issues/643) for which we had to pay twice because it was not correctly implemented for the 1st time. [Here](https://github.com/ubiquity/ubiquity-dollar/issues/643#issuecomment-1607375963) is how it happened. While proper QA eliminates such cases we should also add a penalty for a bounty hunter.

What should be done:
- if issue is reopened and permit was earlier generated then subtract "penalized" issue reward from the next bounty hunter's solved issue (if there is one)
- the bot should post a "penalty" message before generating a permit URL so that a bounty hunter knew that there is a penalty and the reward is reduced

**Comment by pavlovcik** [comment]
Could you propose architecture for this?

For example, at the time of new payment permit generation, we need to check: 
1. the issues that bounty hunter is assigned to 
2. If any of those have been reopened 
3. The permit amount 

If we communicate the debit ONLY at the time of payment permit generation, due to mismatched expectations (they see the bounty price being much higher than what they actually receive) the bounty hunter can be upset/disappointed and quit. 

Is there a clearer way to manage expectations before they get to complete the second bounty?

**Comment by rndquu** [comment]
> Could you propose architecture for this?
> 
> For example, at the time of new payment permit generation, we need to check:
> 
> 1. the issues that bounty hunter is assigned to (what if they un assign themselves after it's been reopened?)
> 2. If any of those have been reopened
> 3. The permit amount
> 
> If we communicate the debit ONLY at the time of payment permit generation, due to mismatched expectations (they see the bounty price being much higher than what they actually receive) the bounty hunter can be upset/disappointed and quit.
> 
> Is there a clearer way to manage expectations before they get to complete the second bounty?

I actually thought that this "penalty" data could be written to a DB. This way we don't need to parse all the issues on permit generation.

So if an issue is reopened then:
1. Find permit comment
2. Find the latest assignment before the permit generation
3. Save bounty hunter's id, penalty amount, issue id
4. On some other permit generation apply the penalty

This schema is cross-repo so it is possible to get a penalty in one repo but have it applied in another one.

> Is there a clearer way to manage expectations before they get to complete the second bounty?

We can soften the penalty a little bit, for example by applying a rule "penalize not more than 50% of a bounty reward"

> the bounty hunter can be upset/disappointed

And we are upset because the previously implemented issue didn't match the spec

**Comment by pavlovcik** [comment]
> And we are upset because the previously implemented issue didn't match the spec

lol

Here's an idea for a new flow:

1. An issue is re-opened due to unsatisfactory work. 
2. At this moment, if the bounty hunter is working on another issue, the bot will immediately post a warning comment. For example: 
    - "@user a prior task you worked on, **#9876**, has been re-opened. Please be sure to review its conversation and implement any necessary fixes. Until that issue is closed again, a penalty of that issue amount will be included in your next payment."

**Comment by rndquu** [comment]
> I kind of like the idea of isolating repositories (or at least organizations) for this because what if the review board of one repository (or organization) is not competent?

Ok, so we need to keep a penalty in boundaries of a single repo.

> At this moment, if the bounty hunter is working on another issue, the bot will immediately post a warning comment

We could post this penalty warning message in the reopened issue. So if issue is reopened then post smth like:
- "@user this task has been reopened. Please be sure to review its conversation and implement any necessary fixes. Until this issue is closed again, a penalty of this issue amount will be included in your next payment."

**Comment by pavlovcik** [comment]
"@user please be sure to review this conversation and implement any necessary fixes. Unless this is closed as completed, its reward of **XXX USD** will be debited from your next payment."

**Comment by rndquu** [comment]
> "@user please be sure to review this conversation and implement any necessary fixes. Unless this is closed as completed, its payment of **XXX USD** will be debited from your next bounty."

Updated description

**Comment by pavlovcik** [comment]
If the second bounty is smaller than the first bounty we won't be made whole by the way.

**Comment by rndquu** [comment]
> If the second bounty is smaller than the first bounty we won't be made whole by the way

Then we will be made whole on the next bounty

> This approach is "optimistic" but if we do "pessimistic" then we can automatically issue only e.g. half the bounty up front and if its not re-opened within a e.g. week then issue the second half payment permit.

This complicates payouts because we would somehow need to maintain 2 unique nonces per a single issue. Anyway proper QA eliminates most of poor PRs so I would not split payouts.

**Comment by Venoox** [comment]
What about a situation where an issue is reopened but in the mean time the bounty hunter completes another issue and gets a penalty but after that fixes the reopened issue.

I guess the warning solves this to some extent by warning the user not to finish other bounties before fixing the reopened issue.

**Comment by rndquu** [comment]
> where an issue is reopened but in the mean time the bounty hunter completes another issue and gets a penalty but after that fixes the reopened issue

Good catch. So if a bounty hunter solves another issue and the penalty is applied then there is no incentive for a bounty hunter to solve a reopened issue.

Perhaps we should apply a progressive penalty rate (like -10% reward for each reopened issue) on all solved bounties in the future until all reopened issues are solved. 

@pavlovcik what do you think?

**Comment by Venoox** [comment]
Also what if the payment token has changed? How do we calculate the penalty on new permit generation with new token?

**Comment by rndquu** [comment]
> Also what if the payment token has changed? How do we calculate the penalty on new permit generation with new token?

I guess payment tokens are always(?) USD pegged so it doesn't really matter which token to use for penalty calculation, DAI or WXDAI.

**Comment by Venoox** [comment]
So the penalty should be written in database and restricted to one repo?

**Comment by rndquu** [comment]
> So the penalty should be written in database

yes, unless there is a way to save it somehow in github comments

> restricted to one repo?

yes

**Comment by pavlovcik** [comment]
> > where an issue is reopened but in the mean time the bounty hunter completes another issue and gets a penalty but after that fixes the reopened issue
> 
> Good catch. So if a bounty hunter solves another issue and the penalty is applied then there is no incentive for a bounty hunter to solve a reopened issue.

Economically, I think that the partner is made whole (they are paid back for non-performance) so it doesn't matter. Let's continue with the original specification.

**Comment by pavlovcik** [comment]
The delays are on us. We should help get @Venoox work merged in asap can you guys help @ubiquity/development

**Comment by rndquu** [comment]
> @ubiquity/software-development has anybody seen this debt system work in production? I have not once seen it.

I haven't seen such cases yet. Perhaps that's because the bot's core team is fixing things all the time without reopening issues because it is simpler to fix the bug themselves compared to finding the root cause PR and waiting for a fix from a bounty hunter.

**Comment by pavlovcik** [comment]
Issues can certainly be opened retroactively. I've done it a few times. 

I suspect that the debt system doesn't work because in our current implementation: even if worker A had their issue re-opened, and B fixed it; A should still have debt for their next bounty.

**Comment by pavlovcik** [comment]
- Found one https://github.com/ubiquity/ubiquibot/issues/680#issuecomment-1715253580

**Comment by 0xcodercrane** [comment]
> @ubiquity/software-development has anybody seen this debt system work in production? I have not once seen it.

yes during the test in ubiquibot/production
